File: Algorithms/20. Valid Parentheses.ts
function isValid(s: string): boolean {
  const left = ["(", "{", "["];
  const stack: string[] = [];
  for (const char of s) {
    if (left.includes(char)) {
      stack.push(char);
    } else {
      if (
        (char === ")" && stack[stack.length - 1] === "(") ||
        (char === "}" && stack[stack.length - 1] === "{") ||
        (char === "]" && stack[stack.length - 1] === "[")
      ) {
        stack.pop();
      } else {
        return false;
      }
    }
  }

  return stack.length === 0;
}

File: Algorithms/20. Valid Parentheses.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        if s.len() == 0 {
            return true;
        }
        if s.len() % 2 == 1 {
            return false;
        }
        let chars: Vec<char> = s.chars().collect();
        let mut stack: Vec<char> = Vec::new();
        // print!("{} {}", chars[0], chars[1]);
        for i in 0..s.len() {
            // let temp = chars[i].to_string();
            match chars[i] {
                '{' => stack.push('{'),
                '(' => stack.push('('),
                '[' => stack.push('['),
                '}' => {
                    if stack.pop() == Some('{') {
                        continue;
                    } else {
                        return false;
                    }
                }
                ')' => {
                    if stack.pop() == Some('(') {
                        continue;
                    } else {
                        return false;
                    }
                }
                ']' => {
                    if stack.pop() == Some('[') {
                        continue;
                    } else {
                        return false;
                    }
                }
                _ => continue,
            }
        }
        stack.len() == 0
    }
}
